File: backend/core/communication_logger.py
"""
Communication logger for tracking all inter-service communication.

Provides structured logging for WebSocket messages, command requests/responses,
and data flows between backend, frontend, and agent services.
"""

from __future__ import annotations

import json
import logging
import os
from datetime import datetime, timezone
from pathlib import Path
from typing import Any, Dict, Optional, Union
import structlog

from backend.core.config import settings

# Communication log file path
PROJECT_ROOT = Path(__file__).resolve().parents[2]
LOGS_ROOT = PROJECT_ROOT / "logs"
COMMUNICATION_LOG_FILE = LOGS_ROOT / "backend" / "communication.log"

# Default configuration values
DEFAULT_MAX_PAYLOAD_SIZE = 10 * 1024  # 10KB
DEFAULT_SENSITIVE_FIELDS = {"password", "token", "api_key", "secret"}

_logger: Optional[logging.Logger] = None
_struct_logger: Optional[structlog.BoundLoggerBase] = None


def _get_config_value(name: str, default: Any) -> Any:
    """Get configuration value from settings or use default."""
    return getattr(settings, name, default)


def _sanitize_payload(payload: Any, sensitive_fields: set = None) -> Any:
    """Sanitize payload by removing or masking sensitive fields."""
    if sensitive_fields is None:
        sensitive_fields = _get_config_value("COMMUNICATION_SENSITIVE_FIELDS", DEFAULT_SENSITIVE_FIELDS)

    if isinstance(payload, dict):
        sanitized = {}
        for key, value in payload.items():
            if key.lower() in sensitive_fields:
                sanitized[key] = "***REDACTED***"
            else:
                sanitized[key] = _sanitize_payload(value, sensitive_fields)
        return sanitized
    elif isinstance(payload, list):
        return [_sanitize_payload(item, sensitive_fields) for item in payload]
    else:
        return payload


def _truncate_payload(payload: Any, max_size: int = None) -> Any:
    """Truncate payload if it exceeds maximum size."""
    if max_size is None:
        max_size = _get_config_value("COMMUNICATION_MAX_PAYLOAD_SIZE", DEFAULT_MAX_PAYLOAD_SIZE)

    payload_str = json.dumps(payload, default=str)
    if len(payload_str) > max_size:
        # Truncate and add indicator
        truncated = payload_str[:max_size - 50] + "...[TRUNCATED]"
        return {"truncated": True, "size": len(payload_str), "content": truncated}

    return payload


def _get_payload_summary(payload: Any) -> Dict[str, Any]:
    """Generate payload summary for logging."""
    if payload is None:
        return {"type": "null", "size_bytes": 0}

    sanitized = _sanitize_payload(payload)
    truncated = _truncate_payload(sanitized)

    summary = {
        "type": type(payload).__name__,
        "size_bytes": len(json.dumps(payload, default=str))
    }

    # Add structure information for complex types
    if isinstance(payload, dict):
        summary["keys"] = list(payload.keys())
        summary["key_count"] = len(payload)
    elif isinstance(payload, list):
        summary["length"] = len(payload)
        if payload and isinstance(payload[0], dict):
            summary["item_keys"] = list(payload[0].keys()) if payload[0] else []

    summary["payload"] = truncated
    return summary


def _ensure_log_directory() -> None:
    """Ensure communication log directory exists."""
    COMMUNICATION_LOG_FILE.parent.mkdir(parents=True, exist_ok=True)


def _get_logger() -> structlog.BoundLoggerBase:
    """Get or create the communication logger."""
    global _logger, _struct_logger

    if _struct_logger is not None:
        return _struct_logger

    # Create dedicated logger for communication
    _logger = logging.getLogger("communication")
    _logger.setLevel(logging.INFO)

    # Create log directory
    _ensure_log_directory()

    # Add file handler for communication logs
    from logging.handlers import RotatingFileHandler
    file_handler = RotatingFileHandler(
        COMMUNICATION_LOG_FILE,
        maxBytes=10 * 1024 * 1024,  # 10MB
        backupCount=5,
        encoding="utf-8"
    )

    # Use JSON formatter for structured logging
    formatter = logging.Formatter(
        '{"timestamp": "%(asctime)s", "level": "%(levelname)s", "message": %(message)s}',
        datefmt="%Y-%m-%dT%H:%M:%SZ"
    )
    formatter.converter = lambda x: datetime.fromtimestamp(x, tz=timezone.utc).timetuple()
    file_handler.setFormatter(formatter)

    _logger.addHandler(file_handler)
    _logger.propagate = False  # Don't propagate to root logger

    # Use structlog.wrap_logger to create a bound logger without overwriting global config
    _struct_logger = structlog.wrap_logger(
        _logger,
        processors=[
            structlog.processors.TimeStamper(fmt="iso", utc=True),
            structlog.processors.JSONRenderer(),
        ],
    )
    return _struct_logger


def log_communication(
    direction: str,  # "inbound" | "outbound"
    protocol: str,  # "websocket" | "redis" | "http"
    message_type: str,  # "data_update" | "command" | "response" | "event"
    resource: Optional[str] = None,  # "signal" | "portfolio" | "trade" | etc.
    correlation_id: Optional[str] = None,
    target: Optional[str] = None,  # "frontend" | "agent" | "backend"
    connection_id: Optional[str] = None,
    payload: Any = None,
    latency_ms: Optional[float] = None,
    error: Optional[str] = None,
    **extra_fields
) -> None:
    """Log inter-service communication event.

    Args:
        direction: Direction of communication ("inbound" | "outbound")
        protocol: Communication protocol ("websocket" | "redis" | "http")
        message_type: Type of message ("data_update" | "command" | "response" | "event")
        resource: Resource type (e.g., "signal", "portfolio", "trade")
        correlation_id: Correlation ID for request/response tracking
        target: Target service ("frontend" | "agent" | "backend")
        connection_id: Connection identifier
        payload: Message payload to log
        latency_ms: Latency in milliseconds
        error: Error message if applicable
        **extra_fields: Additional fields to include in log
    """
    # Check if communication logging is enabled
    if not _get_config_value("ENABLE_COMMUNICATION_LOGGING", True):
        return

    logger = _get_logger()

    # Build log entry
    log_entry = {
        "service": "backend",
        "direction": direction,
        "protocol": protocol,
        "message_type": message_type,
        "resource": resource,
        "correlation_id": correlation_id,
        "target": target,
        "connection_id": connection_id,
    }

    # Add optional fields
    if latency_ms is not None:
        log_entry["latency_ms"] = round(latency_ms, 2)

    if error:
        log_entry["error"] = error

    # Add payload summary if payload provided
    if payload is not None:
        log_entry["payload_summary"] = _get_payload_summary(payload)

    # Add extra fields
    log_entry.update(extra_fields)

    # Log the entry
    logger.info("communication_event", **log_entry)


def log_websocket_message(
    direction: str,
    message_type: str,
    resource: Optional[str] = None,
    correlation_id: Optional[str] = None,
    connection_id: Optional[str] = None,
    payload: Any = None,
    target: str = "frontend"
) -> None:
    """Log WebSocket message communication."""
    log_communication(
        direction=direction,
        protocol="websocket",
        message_type=message_type,
        resource=resource,
        correlation_id=correlation_id,
        connection_id=connection_id,
        payload=payload,
        target=target
    )


def log_agent_command(
    direction: str,
    command: str,
    correlation_id: str,
    payload: Any = None,
    latency_ms: Optional[float] = None,
    error: Optional[str] = None
) -> None:
    """Log agent command communication."""
    log_communication(
        direction=direction,
        protocol="websocket",  # Primary protocol, fallback to redis
        message_type="command" if direction == "outbound" else "response",
        resource=command,
        correlation_id=correlation_id,
        target="agent",
        payload=payload,
        latency_ms=latency_ms,
        error=error
    )


def log_frontend_command(
    direction: str,
    command: str,
    correlation_id: str,
    payload: Any = None,
    latency_ms: Optional[float] = None,
    error: Optional[str] = None
) -> None:
    """Log frontend command communication."""
    log_communication(
        direction=direction,
        protocol="websocket",
        message_type="command" if direction == "inbound" else "response",
        resource=command,
        correlation_id=correlation_id,
        target="frontend",
        payload=payload,
        latency_ms=latency_ms,
        error=error
    )


def log_data_flow(
    source: str,
    destination: str,
    data_type: str,
    data_id: Optional[str] = None,
    payload: Any = None,
    correlation_id: Optional[str] = None
) -> None:
    """Log data flow between services."""
    log_communication(
        direction="outbound",
        protocol="websocket",
        message_type="data_flow",
        resource=data_type,
        correlation_id=correlation_id,
        target=destination,
        payload=payload,
        source=source,
        data_id=data_id
    )


# Utility functions for correlation ID generation
def generate_correlation_id(prefix: str = "corr") -> str:
    """Generate a unique correlation ID."""
    import uuid
    return f"{prefix}_{uuid.uuid4().hex[:16]}"


def extract_correlation_id(message: Dict[str, Any]) -> Optional[str]:
    """Extract correlation ID from message."""
    return message.get("correlation_id") or message.get("request_id")

File: frontend/utils/communicationLogger.ts
/**
 * Frontend communication logger utility.
 *
 * Provides structured logging for WebSocket messages, command requests/responses,
 * and data flows between frontend and backend services.
 */

// Configuration
const DEFAULT_MAX_PAYLOAD_SIZE = 10 * 1024; // 10KB
const DEFAULT_SENSITIVE_FIELDS = ['password', 'token', 'api_key', 'secret'];
const STORAGE_KEY = 'trading_agent_communication_logs';
const MAX_STORED_LOGS = 100;

// Types
interface CommunicationLogEntry {
  timestamp: string;
  service: 'frontend';
  direction: 'inbound' | 'outbound';
  protocol: 'websocket';
  message_type: string;
  resource?: string;
  correlation_id?: string;
  target: 'backend';
  payload_summary: {
    type: string;
    size_bytes: number;
    keys?: string[];
    length?: number;
    payload: any;
  };
  latency_ms?: number;
  error?: string;
  [key: string]: any;
}

interface LogConfig {
  enableLogging: boolean;
  enableStorage: boolean;
  maxPayloadSize: number;
  sensitiveFields: string[];
  maxStoredLogs: number;
}

// Global configuration
let config: LogConfig = {
  enableLogging: process.env.NODE_ENV === 'development', // Default to enabled in development
  enableStorage: false, // Disabled by default to avoid storage bloat
  maxPayloadSize: DEFAULT_MAX_PAYLOAD_SIZE,
  sensitiveFields: DEFAULT_SENSITIVE_FIELDS,
  maxStoredLogs: MAX_STORED_LOGS,
};

/**
 * Configure the communication logger
 */
export function configureLogger(newConfig: Partial<LogConfig>): void {
  config = { ...config, ...newConfig };
}

/**
 * Sanitize payload by removing or masking sensitive fields
 */
function sanitizePayload(payload: any, sensitiveFields: string[] = config.sensitiveFields): any {
  if (!payload) return payload;

  if (typeof payload === 'object' && payload !== null) {
    if (Array.isArray(payload)) {
      return payload.map(item => sanitizePayload(item, sensitiveFields));
    }

    const sanitized: any = {};
    for (const [key, value] of Object.entries(payload)) {
      if (sensitiveFields.some(field => key.toLowerCase().includes(field.toLowerCase()))) {
        sanitized[key] = '***REDACTED***';
      } else {
        sanitized[key] = sanitizePayload(value, sensitiveFields);
      }
    }
    return sanitized;
  }

  return payload;
}

/**
 * Truncate payload if it exceeds maximum size
 */
function truncatePayload(payload: any, maxSize: number = config.maxPayloadSize): any {
  if (!payload) return payload;

  const payloadStr = JSON.stringify(payload);
  if (payloadStr.length > maxSize) {
    return {
      truncated: true,
      size: payloadStr.length,
      content: payloadStr.substring(0, maxSize - 50) + '...[TRUNCATED]'
    };
  }

  return payload;
}

/**
 * Generate payload summary for logging
 */
function getPayloadSummary(payload: any): CommunicationLogEntry['payload_summary'] {
  if (payload === null || payload === undefined) {
    return { type: 'null', size_bytes: 0, payload };
  }

  const sanitized = sanitizePayload(payload);
  const truncated = truncatePayload(sanitized);

  const summary: CommunicationLogEntry['payload_summary'] = {
    type: Array.isArray(payload) ? 'array' : typeof payload,
    size_bytes: JSON.stringify(payload).length,
    payload: truncated
  };

  if (typeof payload === 'object' && payload !== null && !Array.isArray(payload)) {
    summary.keys = Object.keys(payload);
    summary.length = summary.keys.length;
  } else if (Array.isArray(payload)) {
    summary.length = payload.length;
    if (payload.length > 0 && typeof payload[0] === 'object') {
      summary.keys = Object.keys(payload[0]);
    }
  }

  return summary;
}

/**
 * Store log entry in browser storage (if enabled)
 */
function storeLogEntry(entry: CommunicationLogEntry): void {
  if (!config.enableStorage) return;

  try {
    const existing = localStorage.getItem(STORAGE_KEY);
    const logs: CommunicationLogEntry[] = existing ? JSON.parse(existing) : [];

    logs.push(entry);

    // Keep only the most recent logs
    if (logs.length > config.maxStoredLogs) {
      logs.splice(0, logs.length - config.maxStoredLogs);
    }

    localStorage.setItem(STORAGE_KEY, JSON.stringify(logs));
  } catch (error) {
    // Silently fail if storage is not available or full
    console.warn('Failed to store communication log:', error);
  }
}

/**
 * Get stored log entries
 */
export function getStoredLogs(): CommunicationLogEntry[] {
  if (!config.enableStorage) return [];

  try {
    const stored = localStorage.getItem(STORAGE_KEY);
    return stored ? JSON.parse(stored) : [];
  } catch (error) {
    console.warn('Failed to retrieve stored logs:', error);
    return [];
  }
}

/**
 * Clear stored log entries
 */
export function clearStoredLogs(): void {
  try {
    localStorage.removeItem(STORAGE_KEY);
  } catch (error) {
    console.warn('Failed to clear stored logs:', error);
  }
}

/**
 * Log a communication event
 */
export function logCommunication(
  direction: 'inbound' | 'outbound',
  messageType: string,
  payload: any,
  options: {
    resource?: string;
    correlationId?: string;
    latencyMs?: number;
    error?: string;
    extra?: Record<string, any>;
  } = {}
): void {
  if (!config.enableLogging) return;

  const entry: CommunicationLogEntry = {
    timestamp: new Date().toISOString(),
    service: 'frontend',
    direction,
    protocol: 'websocket',
    message_type: messageType,
    target: 'backend',
    payload_summary: getPayloadSummary(payload),
    ...options.extra
  };

  if (options.resource) entry.resource = options.resource;
  if (options.correlationId) entry.correlation_id = options.correlationId;
  if (options.latencyMs !== undefined) entry.latency_ms = options.latencyMs;
  if (options.error) entry.error = options.error;

  // Log to console with structured format
  const logLevel = options.error ? 'error' : 'info';
  console[logLevel](`[Communication:${direction.toUpperCase()}]`, {
    type: messageType,
    resource: options.resource,
    correlationId: options.correlationId,
    payload: entry.payload_summary.payload,
    latencyMs: options.latencyMs,
    error: options.error
  });

  // Store in browser storage
  storeLogEntry(entry);
}

/**
 * Log WebSocket message
 */
export function logWebSocketMessage(
  direction: 'inbound' | 'outbound',
  messageType: string,
  payload: any,
  options: {
    resource?: string;
    correlationId?: string;
  } = {}
): void {
  logCommunication(direction, messageType, payload, {
    resource: options.resource,
    correlationId: options.correlationId
  });
}

/**
 * Log command request
 */
export function logCommandRequest(
  command: string,
  parameters: any,
  requestId: string
): void {
  logCommunication('outbound', 'command', parameters, {
    resource: command,
    correlationId: requestId
  });
}

/**
 * Log command response
 */
export function logCommandResponse(
  command: string,
  response: any,
  requestId: string,
  latencyMs?: number,
  error?: string
): void {
  logCommunication('inbound', 'response', response, {
    resource: command,
    correlationId: requestId,
    latencyMs,
    error
  });
}

/**
 * Log subscription event
 */
export function logSubscription(channels: string[]): void {
  logCommunication('outbound', 'subscription', { channels }, {
    extra: { channels }
  });
}

/**
 * Generate correlation ID
 */
export function generateCorrelationId(prefix: string = 'frontend'): string {
  return `${prefix}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;
}

/**
 * Extract correlation ID from message
 */
export function extractCorrelationId(message: any): string | undefined {
  return message?.correlation_id || message?.request_id;
}

/**
 * Utility to measure latency
 */
export class LatencyTimer {
  private startTime: number;
  private requestId: string;

  constructor(requestId: string) {
    this.startTime = performance.now();
    this.requestId = requestId;
  }

  getElapsedMs(): number {
    return performance.now() - this.startTime;
  }

  end(): number {
    return this.getElapsedMs();
  }
}

// Export types for use in other files
export type { CommunicationLogEntry, LogConfig };